File: buildy_lib/build_state.py
"""
Persistent build state management for incremental builds.
"""

import json
import hashlib
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Dict, Optional, Any
import logging

logger = logging.getLogger(__name__)


@dataclass
class TaskResult:
    """Result of a completed task"""
    task_id: str
    cache_key: str
    success: bool
    outputs: list[str]
    timestamp: float


@dataclass
class BuildState:
    """Persistent build state across invocations"""
    last_build_time: float
    task_graph_hash: str  # Hash of the config + toolchain
    completed_tasks: Dict[str, Dict[str, Any]]  # task_id -> result dict
    file_mtimes: Dict[str, float]  # file_path -> modification time
    config_hash: str  # Hash of build configuration
    toolchain_hash: str = ""  # Hash of toolchain configuration
    
    def save(self, state_file: Path):
        """Save build state to disk"""
        try:
            state_file.parent.mkdir(parents=True, exist_ok=True)
            with open(state_file, 'w') as f:
                json.dump(asdict(self), f, indent=2)
            logger.debug(f"Saved build state to {state_file}")
        except Exception as e:
            logger.warning(f"Failed to save build state: {e}")
    
    @staticmethod
    def load(state_file: Path) -> Optional['BuildState']:
        """Load build state from disk"""
        if not state_file.exists():
            logger.debug(f"No build state found at {state_file}")
            return None
        
        try:
            with open(state_file, 'r') as f:
                data = json.load(f)
            
            # Handle missing fields for backward compatibility
            if 'toolchain_hash' not in data:
                data['toolchain_hash'] = ""
            
            logger.debug(f"Loaded build state from {state_file}")
            return BuildState(**data)
        except Exception as e:
            logger.warning(f"Failed to load build state: {e}")
            return None
    
    @staticmethod
    def hash_config(config: Dict[str, Any]) -> str:
        """Generate hash of configuration"""
        # Convert config to stable JSON string
        config_str = json.dumps(config, sort_keys=True)
        return hashlib.sha256(config_str.encode()).hexdigest()
    
    @staticmethod
    def hash_file(file_path: str) -> str:
        """Generate hash of a file"""
        try:
            with open(file_path, 'rb') as f:
                return hashlib.sha256(f.read()).hexdigest()
        except Exception as e:
            logger.warning(f"Failed to hash file {file_path}: {e}")
            return ""

File: src/internal/cache/build_state.go
package cache

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
)

// TaskResult represents the result of a completed task
type TaskResult struct {
	TaskID    string   `json:"task_id"`
	CacheKey  string   `json:"cache_key"`
	Success   bool     `json:"success"`
	Outputs   []string `json:"outputs"`
	Timestamp float64  `json:"timestamp"`
}

// BuildState represents persistent build state across invocations
type BuildState struct {
	LastBuildTime  float64                    `json:"last_build_time"`
	TaskGraphHash  string                     `json:"task_graph_hash"`  // Hash of the config + toolchain
	CompletedTasks map[string]map[string]interface{} `json:"completed_tasks"` // task_id -> result dict
	FileMtimes     map[string]float64         `json:"file_mtimes"`      // file_path -> modification time
	ConfigHash     string                     `json:"config_hash"`      // Hash of build configuration
	ToolchainHash  string                     `json:"toolchain_hash"`   // Hash of toolchain configuration
}

// NewBuildState creates a new BuildState with default values
func NewBuildState() *BuildState {
	return &BuildState{
		LastBuildTime:  0,
		TaskGraphHash:  "",
		CompletedTasks: make(map[string]map[string]interface{}),
		FileMtimes:     make(map[string]float64),
		ConfigHash:     "",
		ToolchainHash:  "",
	}
}

// Save saves build state to disk
func (bs *BuildState) Save(stateFile string) error {
	// Ensure parent directory exists
	dir := filepath.Dir(stateFile)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("failed to create state directory: %w", err)
	}
	
	// Marshal to JSON
	data, err := json.MarshalIndent(bs, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal build state: %w", err)
	}
	
	// Write to file
	if err := os.WriteFile(stateFile, data, 0644); err != nil {
		return fmt.Errorf("failed to write build state: %w", err)
	}
	
	log.Printf("Saved build state to %s", stateFile)
	return nil
}

// LoadBuildState loads build state from disk
func LoadBuildState(stateFile string) (*BuildState, error) {
	if _, err := os.Stat(stateFile); os.IsNotExist(err) {
		log.Printf("No build state found at %s", stateFile)
		return nil, nil
	}
	
	data, err := os.ReadFile(stateFile)
	if err != nil {
		return nil, fmt.Errorf("failed to read build state: %w", err)
	}
	
	var state BuildState
	if err := json.Unmarshal(data, &state); err != nil {
		return nil, fmt.Errorf("failed to parse build state: %w", err)
	}
	
	// Handle missing fields for backward compatibility
	if state.CompletedTasks == nil {
		state.CompletedTasks = make(map[string]map[string]interface{})
	}
	if state.FileMtimes == nil {
		state.FileMtimes = make(map[string]float64)
	}
	
	log.Printf("Loaded build state from %s", stateFile)
	return &state, nil
}

// HashConfig generates hash of configuration
func HashConfig(config map[string]interface{}) (string, error) {
	// Convert config to stable JSON string (sorted keys)
	configJSON, err := json.Marshal(config)
	if err != nil {
		return "", fmt.Errorf("failed to marshal config: %w", err)
	}
	
	hash := sha256.Sum256(configJSON)
	return fmt.Sprintf("%x", hash[:]), nil
}

// HashFile generates hash of a file
func HashFile(filePath string) (string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return "", fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()
	
	hash := sha256.New()
	if _, err := io.Copy(hash, file); err != nil {
		return "", fmt.Errorf("failed to hash file: %w", err)
	}
	
	return fmt.Sprintf("%x", hash.Sum(nil)), nil
}

